### PR TITLE
feat(evo2): Parallel evo2 batched generation

### PR DESCRIPTION
**Pull request by jstjohn**
### Description

Splitting the batched inference and inference speed improvements out of https://github.com/NVIDIA-BioNeMo/bionemo-recipes/pull/1699. This adds the ability to do parallel batched generation, and bumps the versions of mbridge and megatron.

#### Usage

```
infer_evo2 --help
```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests, including unit tests, slow tests, notebooks, and every recipe/model directory.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

#### Triggering Code Rabbit AI Review

To trigger a code review from code rabbit, comment on a pull request with one of these commands:

- @coderabbitai review - Triggers a standard review
- @coderabbitai full review - Triggers a comprehensive review

See https://docs.coderabbit.ai/reference/review-commands for a full list of commands.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

* **New Features**
  * Added optional batched decoding for same-length prompts, with streaming output and richer JSONL results.
  * Added controls to ignore end-of-sequence tokens and enforce strict generation validation.
  * Added `--no-save-optim` to exclude optimizer state from training checkpoints.
  * Added optional timing and memory metrics for inference runs.

* **Bug Fixes**
  * Improved inference stability, stop-token handling, dynamic state management, and CUDA graph behavior.
  * Updated guest access configuration for more reliable data downloads.

* **Performance**
  * Updated Evo2 components and improved memory sizing for hybrid sequence processing.

### CODE REVIEW

File: recipes/evo2_megatron/src/bionemo/evo2/run/infer.py [hunk 1]
```diff
@@ -1626,55 +2317,111 @@ def infer(
     total_prompt_tokens = 0
     total_completion_tokens = 0
     t_generate_start = time.perf_counter()
+    is_rank_zero = int(os.environ.get("RANK", "0")) == 0
```
Comment (farhadrgh) [inline recipes/evo2_megatron/src/bionemo/evo2/run/infer.py]: The old code used `parallel_state.get_data_parallel_rank() == 0`   
The current version is `int(os.environ.get("RANK", "0")) == 0`. I think these are different in TP runs: with TP=4 and no pipeline/data parallelism, all four ranks have `get_data_parallel_rank() == 0`, so the old code would have every TP rank write the output file, which is almost certainly a bug. The new env-var approach correctly restricts I/O to global rank 0.

If this is an intentional fix, a one-line comment would help future readers understand why the env var is preferred over the parallel_state API here.

File: recipes/evo2_megatron/src/bionemo/evo2/run/infer.py [hunk 2]
```diff
@@ -813,13 +910,147 @@ def _modify_for_top_p(logits, p):
...
+    Individually added requests receive fresh Mamba slots in descending order. Before any
+    recurrent state is consumed, reverse that exact allocator order in the request mapping so
+    request rows align with the ascending stable slice used by packed Hyena state. Leave every
+    other order unchanged so the binding validation rejects arbitrary permutations.
+    """
+    request_slots = dyn_ctx.mamba_metadata.request_to_mamba_state_idx[:request_count]
+    slots = [int(slot) for slot in request_slots.tolist()]
+    expected_lifo_slots = list(range(slots[0], slots[0] - len(slots), -1)) if slots else []
+    if slots and slots == expected_lifo_slots:
+        request_slots.copy_(request_slots.flip(0))
```
Comment (farhadrgh) [inline recipes/evo2_megatron/src/bionemo/evo2/run/infer.py:1057]: This mutates dyn_ctx.mamba_metadata.request_to_mamba_state_idx in-place. Does MCore read this array again after bind_hyena_packed_views_to_dynamic_context_batch completes (e.g. during update_requests or initialize_attention_state)? If so, MCore would see the reversed (ascending) order it didn't allocate, which could silently corrupt state.

**Review by farhadrgh — APPROVED** [review]
LGTM

File: recipes/evo2_megatron/src/bionemo/evo2/run/infer.py (hunk 2 above)
Comment (jstjohn) [reply recipes/evo2_megatron/src/bionemo/evo2/run/infer.py:1057]: I'll look into that, good eye!

File: recipes/evo2_megatron/src/bionemo/evo2/run/infer.py (hunk 1 above)
Comment (jstjohn) [reply recipes/evo2_megatron/src/bionemo/evo2/run/infer.py]: I think API state would be best here. I'll make that change.

**Review by pstjohn — APPROVED** [review]
